### PR TITLE
Isolate each test in a parentless module.

### PR DESCRIPTION
**Pull request by yuyichao**
Replaces https://github.com/JuliaLang/julia/pull/9798.
This uses a parentless module (similar to https://github.com/JuliaLang/julia/pull/9798) since I was interested in the memory saving (not much but I haven't measured too carefully).

There are two test breakage that I need to workaround (instead of fixing).
1. Defining a new module on one process is not visible on another one, which breaks the `DictChannel` test in `examples`.
2. The dictionary key completion doesn't work for qualified names.

There might be a way to fix the first one (or just disable isolation for it as what I'm doing now is probably good enough). The second one is likely a bug that should be fixed.

### CODE REVIEW

**Comment by yuyichao** [comment]
I pushed a change to make the example test pass while being isolated by running the remote on the same worker. I'm not sure if it is still testing everything needed though. @amitmurthy

**Comment by tkelman** [comment]
Was it @dhoegh who implemented the dict key completion? Pretty recently iirc.

**Comment by yuyichao** [comment]
It looks reasonably easy to fix. I'll try later.

**Comment by dhoegh** [comment]
Nope, but a sensible guess. It was @bjarthur.

**Comment by tkelman** [comment]
sorry, should have actually checked blame!

**Comment by yuyichao** [comment]
Dict completion fixed. Rebased to the first commit so it can be merged independently if necessary.

**Comment by amitmurthy** [comment]
I think it is important to actually have the example test run with a remote worker. That test has also uncovered an as yet open issue - https://github.com/JuliaLang/julia/issues/16091

File: test/replcompletions.jl
Comment (tkelman) [inline test/replcompletions.jl]: if this doesn't take too long to run, should we make it a loop over both the `CompletionFoo.test_dict` version and the no-module `test_dict` version?

File: base/REPLCompletions.jl
Comment (tkelman) [inline base/REPLCompletions.jl]: why the change from the symbol to `true` ?

Comment (yuyichao) [inline base/REPLCompletions.jl]: it's not used.

File: base/replutil.jl
Comment (tkelman) [inline base/replutil.jl]: is this an independent bugfix that should have its own added tests?

Comment (yuyichao) [inline base/replutil.jl]: It is caught by wrapping the test in module.

File: test/replcompletions.jl
Comment (yuyichao) [inline test/replcompletions.jl]: The same applies to almost all other tests in the same file though (this was basically the only one that uses a non-qualified name in the replcompletion test before). It's mainly to avoid polluting the namespace but making the name slightly longer could also be fine.

File: test/testdefs.jl
Comment (tkelman) [inline test/testdefs.jl]: this form of Module constructor is apparently undocumented

File: test/replcompletions.jl
Comment (tkelman) [inline test/replcompletions.jl]: > this was basically the only one that uses a non-qualified name in the replcompletion test before

Then shouldn't this be kept for the sake of test coverage?

File: test/testdefs.jl
Comment (yuyichao) [inline test/testdefs.jl]: It's basically an internal function (Although PyCall uses it [last time I checked](https://github.com/JuliaLang/julia/issues/11395)). I'm trying to measure the difference in memory usage now. If there isn't a big difference (likely) I'll probably switch to the non-parentless version below.

File: test/replcompletions.jl
Comment (yuyichao) [inline test/replcompletions.jl]: Well, I think it makes sense to add it back with a longer name but no it won't help test coverage since this one uses a completely different code path that isn't shared with anything else.

Comment (tkelman) [inline test/replcompletions.jl]: module qualified vs not qualified by a module. it seems like you're no longer testing the unqualified version?

**Comment by tkelman** [comment]
I'd like to do the first commit before branching but hold off on the second until after.

**Comment by yuyichao** [comment]
> I'd like to do the first commit before branching (it's a bugfix) but hold off on the second until after.

Sure. FWIW the kwarg error one is probably more important but there isn't a test for it without the second commit now. I'll split this two out into a separate PR shortly.

File: test/runtests.jl
Comment (amitmurthy) [inline test/runtests.jl]: One option is to split `examples.jl` and move the clustermanager and dictchannel tests to `parallel_exec.jl`. With an appropriate comment in `examples.jl` explaining the rationale for the move.

**Comment by yuyichao** [comment]
Still waiting on branching and the appveyor failure is the git network issue...

I switch to normal module now since anonymous modules are going to be kept alive by the types defined in it anyway....

File: test/misc.jl
Comment (tkelman) [inline test/misc.jl]: typo, methodswith ?

**Comment by yuyichao** [comment]
Comment addressed local tests passed and CI re-enabled. This should be ready to go.

I've also tested locally that turning off isolation still pass tests and running all the tests on a single process also works without additional warnings.

All the test files can be run with both mode now. A few things that needs to be run on `Main` are explicitly `eval`d in `Main` (e.g. the REPL, the serializer and the `TestHelper` module). I've also removed a few modules in the dates tests since they are not needed anymore.

**Review by vtjnash — APPROVED** [review]

**Comment by yuyichao** [comment]
Failure was https://github.com/JuliaLang/julia/issues/19132 rebased and will merge when CI passed.

**Pull request merged**
